File: 0079-word-search/solution.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
 
var _exist = function (board, word, i, j, k, arr) {

    if (arr.some(a => a.i === i && a.j === j)) return false;
    arr.push({i, j});

    if (k === word.length) return true;

    if (i > 0 && board[i-1][j] === word[k]) {
        if (_exist(board, word, i-1, j, k+1, arr))
            return true;
    }

    if (i < board.length-1 && board[i+1][j] === word[k]) {
        if (_exist(board, word, i+1, j, k+1, arr))
            return true;
    }

    if (j > 0 && board[i][j-1] === word[k]) {
        if (_exist(board, word, i, j-1, k+1, arr))
            return true;
    }

    if (j < board[i].length-1 && board[i][j+1] === word[k]) {
        if (_exist(board, word, i, j+1, k+1, arr))
            return true;
    }
    arr.pop();
    return false;
}

var exist = function(board, word) {
    for(let i = 0; i < board.length; i++) {
        for(let j = 0; j < board[i].length; j++) {
            if (board[i][j] === word[0]) {
                if(_exist(board, word, i, j, 1, []))
                    return true;
            }
        }
    }
    return false;
};

File: 0079-word-search/solution.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows, cols, path = len(board), len(board[0]), set()

        def dfs(r, c, i):
            if i == len(word): return True
            if (not r in range(0, rows) or
                not c in range(0, cols) or
                word[i] != board[r][c] or
                (r, c) in path): return False
            path.add((r, c))
            if dfs(r + 1, c, i + 1): return True
            if dfs(r - 1, c, i + 1): return True
            if dfs(r, c + 1, i + 1): return True
            if dfs(r, c - 1, i + 1): return True
            path.remove((r, c))
            return False

        for r in range(rows):
            for c in range(cols):
                if dfs(r, c, 0): return True

        return False


